File: chat-service/app/services/chat_manager.py
import asyncio
import json
from collections import defaultdict
from dataclasses import dataclass
from datetime import datetime, timezone
from fastapi import WebSocket
from app.core.config import get_settings
from app.db.database import get_db, get_redis


@dataclass(frozen=True)
class SocketClient:
    websocket: WebSocket
    role: str
    actor: str


class ChatManager:
    def __init__(self) -> None:
        self._rooms: dict[int, set[SocketClient]] = defaultdict(set)
        self._lock = asyncio.Lock()

    async def connect(self, lead_id: int, websocket: WebSocket, role: str, actor: str) -> None:
        await websocket.accept()
        client = SocketClient(websocket=websocket, role=role, actor=actor)
        async with self._lock:
            self._rooms[lead_id].add(client)

        redis = await get_redis()
        await redis.sadd("chat:active_leads", str(lead_id))

        await self.broadcast_event(
            lead_id,
            {
                "event": "presence",
                "lead_id": lead_id,
                "role": role,
                "actor": actor,
                "at": datetime.now(timezone.utc).isoformat(),
            },
            persist=False,
        )

    async def disconnect(self, lead_id: int, websocket: WebSocket) -> None:
        async with self._lock:
            if lead_id not in self._rooms:
                return

            remaining = {
                c for c in self._rooms[lead_id] if c.websocket is not websocket
            }
            if remaining:
                self._rooms[lead_id] = remaining
            else:
                self._rooms.pop(lead_id, None)

        redis = await get_redis()
        room_size = await self.active_count(lead_id)
        if room_size == 0:
            await redis.srem("chat:active_leads", str(lead_id))

    async def active_count(self, lead_id: int) -> int:
        async with self._lock:
            return len(self._rooms.get(lead_id, set()))

    async def broadcast_event(self, lead_id: int, payload: dict, persist: bool = False) -> None:
        if persist:
            await self._save_message(lead_id, payload)

        data = json.dumps(payload)
        async with self._lock:
            clients = list(self._rooms.get(lead_id, set()))

        stale: list[SocketClient] = []
        for client in clients:
            try:
                await client.websocket.send_text(data)
            except Exception:
                stale.append(client)

        if stale:
            async with self._lock:
                for client in stale:
                    self._rooms[lead_id].discard(client)

        redis = await get_redis()
        settings = get_settings()
        await redis.publish(settings.chat_events_channel, data)

    async def _save_message(self, lead_id: int, payload: dict) -> None:
        db = await get_db()
        sender_role = payload.get("sender_role", "system")
        sender_label = payload.get("sender_label", "System")
        message_type = payload.get("message_type", "TEXT")
        content = payload.get("content", payload.get("text", ""))
        metadata = payload.get("metadata", {})

        await db.execute(
            """
            INSERT INTO chat_messages(lead_id, sender_role, sender_label, message_type, content, metadata)
            VALUES($1, $2, $3, $4, $5, $6::jsonb)
            """,
            lead_id,
            sender_role,
            sender_label,
            message_type,
            content,
            json.dumps(metadata),
        )

    async def fetch_history(self, lead_id: int) -> list[dict]:
        db = await get_db()
        rows = await db.fetch(
            """
            SELECT sender_role, sender_label, message_type, content, metadata, created_at
            FROM chat_messages
            WHERE lead_id=$1
            ORDER BY created_at ASC
            """,
            lead_id,
        )

        def normalize_metadata(raw: object) -> dict:
            if raw is None:
                return {}
            if isinstance(raw, dict):
                return raw
            if isinstance(raw, str):
                try:
                    parsed = json.loads(raw)
                except Exception:
                    return {}
                return parsed if isinstance(parsed, dict) else {}
            return {}

        return [
            {
                "event": "message",
                "lead_id": lead_id,
                "sender_role": row["sender_role"],
                "sender_label": row["sender_label"],
                "message_type": row["message_type"],
                "content": row["content"],
                "metadata": normalize_metadata(row["metadata"]),
                "created_at": row["created_at"].isoformat(),
            }
            for row in rows
        ]


chat_manager = ChatManager()

File: chat-service/src/main/java/com/aditistays/chatservice/service/ChatManager.java
package com.aditistays.chatservice.service;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import com.aditistays.chatservice.config.AppProperties;
import com.aditistays.chatservice.repository.ChatRepository;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.SneakyThrows;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.TextMessage;
import org.springframework.web.socket.WebSocketSession;

/** Mirrors chat-service's Python app/services/chat_manager.py. */
@Component
public class ChatManager {

    private static final String ACTIVE_LEADS_SET = "chat:active_leads";

    public record SocketClient(WebSocketSession session, String role, String actor) {
    }

    private final ChatRepository chatRepository;
    private final StringRedisTemplate redisTemplate;
    private final ObjectMapper objectMapper;
    private final String chatEventsChannel;

    private final Map<Long, Set<SocketClient>> rooms = new ConcurrentHashMap<>();
    private final Object lock = new Object();

    public ChatManager(ChatRepository chatRepository, StringRedisTemplate redisTemplate, ObjectMapper objectMapper, AppProperties props) {
        this.chatRepository = chatRepository;
        this.redisTemplate = redisTemplate;
        this.objectMapper = objectMapper;
        this.chatEventsChannel = props.chatEventsChannel();
    }

    public void connect(long leadId, WebSocketSession session, String role, String actor) {
        synchronized (lock) {
            rooms.computeIfAbsent(leadId, k -> ConcurrentHashMap.newKeySet()).add(new SocketClient(session, role, actor));
        }
        redisTemplate.opsForSet().add(ACTIVE_LEADS_SET, String.valueOf(leadId));

        broadcastEvent(leadId, Map.of(
                "event", "presence",
                "lead_id", leadId,
                "role", role,
                "actor", actor,
                "at", Instant.now().toString()
        ), false);
    }

    public void disconnect(long leadId, WebSocketSession session) {
        synchronized (lock) {
            Set<SocketClient> set = rooms.get(leadId);
            if (set != null) {
                set.removeIf(c -> c.session().getId().equals(session.getId()));
                if (set.isEmpty()) {
                    rooms.remove(leadId);
                }
            }
        }
        if (activeCount(leadId) == 0) {
            redisTemplate.opsForSet().remove(ACTIVE_LEADS_SET, String.valueOf(leadId));
        }
    }

    public int activeCount(long leadId) {
        synchronized (lock) {
            Set<SocketClient> set = rooms.get(leadId);
            return set == null ? 0 : set.size();
        }
    }

    @SneakyThrows
    public void broadcastEvent(long leadId, Map<String, Object> payload, boolean persist) {
        if (persist) {
            saveMessage(leadId, payload);
        }

        String data = objectMapper.writeValueAsString(payload);

        List<SocketClient> clients;
        synchronized (lock) {
            clients = new ArrayList<>(rooms.getOrDefault(leadId, Set.of()));
        }

        List<SocketClient> stale = new ArrayList<>();
        for (SocketClient client : clients) {
            try {
                client.session().sendMessage(new TextMessage(data));
            } catch (Exception e) {
                stale.add(client);
            }
        }
        if (!stale.isEmpty()) {
            synchronized (lock) {
                Set<SocketClient> set = rooms.get(leadId);
                if (set != null) {
                    set.removeAll(stale);
                }
            }
        }

        redisTemplate.convertAndSend(chatEventsChannel, data);
    }

    @SuppressWarnings("unchecked")
    private void saveMessage(long leadId, Map<String, Object> payload) {
        String senderRole = String.valueOf(payload.getOrDefault("sender_role", "system"));
        String senderLabel = String.valueOf(payload.getOrDefault("sender_label", "System"));
        String messageType = String.valueOf(payload.getOrDefault("message_type", "TEXT"));
        Object contentObj = payload.containsKey("content") ? payload.get("content") : payload.get("text");
        String content = contentObj == null ? "" : String.valueOf(contentObj);
        Object metadataObj = payload.get("metadata");
        Map<String, Object> metadata = metadataObj instanceof Map ? (Map<String, Object>) metadataObj : Map.of();

        chatRepository.saveMessage(leadId, senderRole, senderLabel, messageType, content, metadata);
    }

    public List<Map<String, Object>> fetchHistory(long leadId) {
        return chatRepository.fetchHistory(leadId);
    }
}
